libraries/AC_CustomControl/AC_CustomControl.cpp

```diff
@@ -0,0 +1,175 @@
+    AP_GROUPINFO("AXIS_MASK", 2, AC_CustomControl, _custom_controller_mask, 0),
+
+    // parameters for empty controller
+    AP_SUBGROUPVARPTR(_backend[0], "1_", 6, AC_CustomControl, _backend_var_info[0]),
+
+    // parameters for PID controller
+    AP_SUBGROUPVARPTR(_backend[1], "2_", 7, AC_CustomControl, _backend_var_info[1]),
+
+    AP_GROUPEND
+};
+
+const struct AP_Param::GroupInfo *AC_CustomControl::_backend_var_info[CUSTOMCONTROL_MAX_TYPES];
+
+AC_CustomControl::AC_CustomControl(AP_AHRS_View*& ahrs, AC_AttitudeControl_Multi*& atti_control, AP_MotorsMulticopter*& motors, float dt) :
+    _dt(dt),
+    _ahrs(ahrs),
+    _atti_control(atti_control),
+    _motors(motors)
+{
+    AP_Param::setup_object_defaults(this, var_info);    
+}
+
+void AC_CustomControl::init(void)
+{
+    switch (CustomControlType(_controller_type))
+    {
+        case CustomControlType::CONT_NONE:
+            break;
+        case CustomControlType::CONT_EMPTY:
+            _backend[get_type()] = new AC_CustomControl_Empty(*this, _ahrs, _atti_control, _motors, _dt);
+            _backend_var_info[get_type()] = AC_CustomControl_Empty::var_info;
+            break;
+        case CustomControlType::CONT_PID:
+            _backend[get_type()] = new AC_CustomControl_PID(*this, _ahrs, _atti_control, _motors, _dt);
+            _backend_var_info[get_type()] = AC_CustomControl_PID::var_info;
+            break;
+        default:
+            return;
+    }
+
+    if (_backend[get_type()] && _backend_var_info[get_type()]) {
+        AP_Param::load_object_from_eeprom(_backend[get_type()], _backend_var_info[get_type()]);
+    }
+}
+
+// run custom controller if it is activated by RC switch and appropriate type is selected
+void AC_CustomControl::update(void)
+{
+    if (is_safe_to_run()) {
+        Vector3f motor_out_rpy;
+
+        motor_out_rpy = _backend[get_type()]->update();   
+
+        motor_set(motor_out_rpy);
+    }
+}
+
+// choose which axis to apply custom controller output
+void AC_CustomControl::motor_set(Vector3f rpy) {
+    if (_custom_controller_mask & (uint8_t)CustomControlOption::ROLL) {
+        _motors->set_roll(rpy.x);
+    }
+    if (_custom_controller_mask & (uint8_t)CustomControlOption::PITCH) {
+        _motors->set_pitch(rpy.y);
+    }
+    if (_custom_controller_mask & (uint8_t)CustomControlOption::YAW) {
+        _motors->set_yaw(rpy.z);
+    }
+}
+
+// move main controller's target to current states, reset filters, 
+// and move integrator to motor output
+// to allow smooth transition to the primary controller
+void AC_CustomControl::reset_main_atti_controller(void) 
+{
+    // reset attitude and rate target, if feedforward is enabled
+    if (_atti_control->get_bf_feedforward()) {
+        _atti_control->relax_attitude_controllers();
+    }
+
+    Vector3f gyro_latest = _ahrs->get_gyro_latest();
+
+    _atti_control->get_rate_roll_pid().set_integrator(_atti_control->rate_bf_targets().x - gyro_latest.x, _motors->get_roll());
```

Comment:
I'm not sure this is a good idea, it might end up setting a massive I term that would take ages to decay off. Might be better to just 0?

Reply:
I agree with your reasoning, I'll do SITL experiment with extreme gains to see its effect. We have multiple examples in POS control library for setting integrator to a non-zero value. But for attitude controller the motors have a much faster response time than the attitude controller itself, so the twitch would probably be very hard to detect with the naked eye. 
https://github.com/ArduPilot/ardupilot/blob/master/libraries/AC_AttitudeControl/AC_PosControl.cpp#L957
https://github.com/ArduPilot/ardupilot/blob/master/libraries/AC_AttitudeControl/AC_PosControl.cpp#L540
https://github.com/ArduPilot/ardupilot/blob/master/libraries/AC_AttitudeControl/AC_PosControl.cpp#L768

Maybe @lthall can help us with this.